### PR TITLE
Start up optimizations grab bag

### PR DESCRIPTION
**Pull request by benthecarman**
#1016 

A bunch of startup optimizations, in my testing this brings down the startup time on a wallet with a lot of channels and a configured fedimint from 12 seconds to 6 seconds.

- 1st is mainly just adding the logs everywhere I have been using. Does an optimization for new wallets, where it'll now create the first node in the background instead of waiting for that to be initialized before returning
- 2nd backgrounds the saving of our node storage on startup, this was only done so we save any changes made to our LSP config while starting up (just migrations), these aren't the end of the world if they are missed, they'll persisted on the next startup 
- 3rd makes it so we don't block on retrying spendable outputs, these can happen async and there isn't a reason to wait
- 4th fixes two issues with fedimint startup. We would block on setting the preferred guardian, this is now done in the background. We also would download the federation info everytime but we only needed it if our db didn't have it (we were already only using the federation info if we didn't have it in our db), this fixes to only do the download if its not in our db. This also does some minor touch ups like initializing federation_id instead of doing federation_info.federation_id() since this would do a hash everytime.
- 5th changes to immediately start fetching the JWT for auth, before we would only fetch it once we got our first VSS request, this saves us about ~200ms because otherwise we wait until we have read all of indexed db until we make this call. Needed to modify the auth client to only allow one call to retrieve_new_jwt at a time, otherwise this would sometimes not complete by the time the first VSS call came in and we'd request multiple JWTs. Otherwise just adds some timing logs and tiny optimizations to indexed db
- 6th changes to sync scorer in the background rather than upfront, this made the startup ~500ms faster for me. This will still read the scorer on start up so we won't be without one until it is synced.

### CODE REVIEW

File: mutiny-core/src/nodemanager.rs [hunk 1]
```diff
@@ -620,11 +619,6 @@ impl<S: MutinyStorage> NodeManager<S> {
     /// Creates a background process that will sync the wallet with the blockchain.
     /// This will also update the fee estimates every 10 minutes.
     pub fn start_sync(nm: Arc<NodeManager<S>>) {
-        // If we are stopped, don't sync
-        if nm.stop.load(Ordering::Relaxed) {
-            return;
-        }
-
```
Comment (benthecarman) [inline mutiny-core/src/nodemanager.rs:585]: removed this because we just immediately do it again in the loop below

File: mutiny-wasm/src/lib.rs [hunk 2]
```diff
@@ -223,6 +223,13 @@ impl MutinyWallet {
                 auth_url,
             ));
 
+            // immediately start fetching JWT
+            let auth = auth_client.clone();
+            spawn(async move {
+                // if this errors, it's okay, we'll call it again when we fetch vss
+                let _ = auth.authenticate().await;
```
Comment (TonyGiorgio) [inline mutiny-wasm/src/lib.rs:230]: maybe a log here would be nice

**Review by TonyGiorgio — COMMENTED** [review]
mostly ack otherwise, just a comment about syncing chain data before nodes

File: mutiny-core/src/lib.rs [hunk 3]
```diff
@@ -810,19 +839,31 @@ impl<S: MutinyStorage> MutinyWalletBuilder<S> {
...
+            // spawn in background, this can take a while and we don't want to block
+            utils::spawn(async move {
+                if let Err(e) = nm.new_node().await {
+                    log_error!(nm.logger, "Failed to create first node: {e}");
+                }
+            })
         };
 
+        // start syncing node manager
+        NodeManager::start_sync(mw.node_manager.clone());
```
[the remark this answers is not in the document]
Comment (benthecarman) [reply mutiny-core/src/lib.rs:853]: I am just moving this call down, we always synced in the background. My understanding is that this is safe and if it is unsafe then we were unsafe before too. Currently we just block on connecting to a peer until our fee estimates are synced, maybe we should be waiting for ldk to sync too.

Asked in discord for confirmation.

[the remark this answers is not in the document]
Comment (benthecarman) [reply mutiny-core/src/lib.rs:853]: It should be okay

image

[the remark this answers is not in the document]
Comment (TonyGiorgio) [reply mutiny-core/src/lib.rs:853]: Where does the fee estimation fit into all of this?

[the remark this answers is not in the document]
Comment (benthecarman) [reply mutiny-core/src/lib.rs:853]: This was my original question. We currently don't connect to a peer until we have fee estimates

image

**Comment by TonyGiorgio** [comment]
Screenshot 2024-02-09 at 4 21 02 PM

Just saw a stream of thousands of duplicate requests getting block height and tip until it panic'd, right after a refresh. Browser completely crashed and couldn't refresh.

**Comment by benthecarman** [comment]
> My on chain balance went away when I refreshed after swapping into a channel. I refreshed again, waited ~30s and then it appeared again. I think chain data needs to be cached or something? Not sure why it went away, never saw this issue before.

This happened to me once when testing the DLC stuff, I think its a bug in bdk with spending unconfirmed utxos. I assume/hope when we update it'll fix

**Comment by benthecarman** [comment]
> Just saw a stream of thousands of duplicate requests getting block height and tip until it panic'd, right after a refresh. Browser completely crashed and couldn't refresh.

do you have any steps to reproduce? syncing logic hasn't changed at all, just starting after we start the nostr stuff instead of before

**Comment by TonyGiorgio** [comment]
> > Just saw a stream of thousands of duplicate requests getting block height and tip until it panic'd, right after a refresh. Browser completely crashed and couldn't refresh.
> 
> do you have any steps to reproduce? syncing logic hasn't changed at all, just starting after we start the nostr stuff instead of before

I just did a lot of refreshes as I went through E2E testing.

**Comment by benthecarman** [comment]
Seems like that is happening during BDK sync, maybe just move it back to where it was until #1002  is merged

**Review by TonyGiorgio — APPROVED** [review]

**Pull request merged**